Documentation/design-docs/framework-version-resolution.md

```diff
@@ -0,0 +1,290 @@
+* Pre-release version will never roll forward to release version. So for example `3.0.0-preview4-27415-15` will not roll forward to `3.0.0`. Also pre-release will only roll forward to the same `major.minor.patch`. So for example `3.0.0-preview4-27415-15` will not roll forward to `3.0.1-preview1-29000-0`.
+
+### New behavior
+The newly proposed behavior is to treat pre-release versions no different from release versions. This would mean that pre-release can roll forward to release but also that release can roll forward to pre-release. For the most part this improves the previous behavior (except when `Disable` is active in which case no roll-forward will happen):
+* `3.0.0-preview4-27415-15` will roll forward to `3.0.0`.
+* `3.0.0-preview4-27415-15` will roll forward to `3.0.1-preview1-29000-0`.
+
+But it also introduces cases which are somewhat controversial:
+* On `Minor` (the default) `3.0.0` would roll forward to `3.0.1-preview1-29000-0` even if `3.0.1` is available on the machine. This behavior is typically not expected.
+* On `LatestMajor` `2.0.0` would roll forward to `3.0.1-preview1-29000-0` even when `3.0.0` is available on the machine.
+
+Pros:
+* Consistent behavior across all versions
+
+Cons:
+* Installing a pre-release version affects apps using release versions
+* Works well only if pre-release versions have similar compatibility behavior as release versions - not many breaking changes. Also expects pre-release versions to generally work well.
+
+### Alternative to new behavior
+When resolving framework reference with a pre-release version, treat all versions the same and include both release and pre-release versions in the set. This means pre-release can resolve to both release or pre-release.
+
+When resolving framework reference with a release version, prefer release versions. This means that if there's a release version on the machine which satisfies all the requirements it will be chosen, regardless of what pre-release versions are installed. Only if there's no suitable release version, then pre-release versions are also considered (and then they're treated all the same).
+
+Given the above examples this would lead to this behavior:
+* On `Minor` (the default) `3.0.0` would roll forward to `3.0.1` even if `3.0.1-preview1-29000-0` is also available on the machine.
+* On `LatestMajor` `2.0.0` would roll forward to `3.0.0` even if `3.0.1-preview1-29000-0` is also available on the machine.
+
+Pros:
+* Installing pre-release version doesn't affect apps which use release version (unless it's needed to make the app work at all)
+* Doesn't impose any implicit expectations on the quality of pre-release versions as typically pre-release would only be used when explicitly asked for.
+* Closer to what most users would expect
+
+Cons:
+* Some special cases won't work
```

Comment:
Any special case that won't work is an example of a case we cannot test until release day or using bits that lie about being rtm.

Reply:
I think this is now resolved with the new version.